### PR TITLE
Automated cherry pick of #108366 (release-1.22): Delay writing a terminal phase until the pod is terminated

### PR DESCRIPTION
**Pull request by bobbypage**
Cherry pick of https://github.com/kubernetes/kubernetes/pull/108366 on release-1.22.

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
Fixed a 1.22 regression that could incorrectly reject pods with OutOfCpu errors if they were rapidly scheduled after other pods were reported as complete in the API. The Kubelet now waits to report the phase of a pod as terminal in the API until all running containers are guaranteed to have stopped and no new containers can be started.  Short-lived pods may take slightly longer (~1s) to report Succeeded or Failed after this change.
```

### CODE REVIEW

**Comment by bobbypage** [comment]
When applying the cherrypick there were four conflicts -

```
+++ Conflicts detected:
UU pkg/kubelet/kubelet.go
UU pkg/kubelet/pod_workers_test.go
UU pkg/kubelet/status/testing/mock_pod_status_provider.go
UU pkg/kubelet/kubelet.go
```

Here are the files with merge markers before I resolved them manually -

`pkg/kubelet/kubelet.go` - https://gist.github.com/d7a1c46beaa3f7a61dff7b5f4c35bbc5
`pkg/kubelet/pod_workers_test.go` - https://gist.github.com/bobbypage/6c057be48caa90d72fbd389fcb12805a
`pkg/kubelet/status/testing/mock_pod_status_provider.go` - https://gist.github.com/ecbdfe8574f490210953d59c00a1c7e4
`test/e2e_node/node_shutdown_linux_test.go` - https://gist.github.com/f94728980d7bb7cbf6eeebddbf84ba60

**Comment by bobbypage** [comment]
[Ran the new e2e serial test manually](https://gist.github.com/94b8d2f2dbdb44698bb306e2ab3f1965) under kind and it passed ([build-log](https://gist.github.com/74b2811bdcbbab97423ba3d565c79556))

**Comment by bobbypage** [comment]
Ran all node e2e serial logs and they passed:

```
$ bash ~/dev_scripts/e2e_node_kubetest_serial.sh # https://gist.github.com/aa658020c05eef18644915dc1cc5afeb
```
[build-log.txt](https://gist.github.com/73b20a799905fac56fa2f7424ba94e6a)

**Comment by bobbypage** [comment]
Removing WIP since [master PR](https://github.com/kubernetes/kubernetes/pull/108366) has merged.

**Comment by liggitt** [comment]
/hold cancel

multiple green runs of post-submit serial test jobs on master with this fix

**Comment by bobbypage** [comment]
Ping release managers for cherrypick approval.

/cc kubernetes/release-managers

also 
/cc @saschagrunert who approved same cherrypick to 1.23 - https://github.com/kubernetes/kubernetes/pull/108723

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Pull request merged**
